### PR TITLE
Aturik/compare frontend

### PR DESCRIPTION
**Pull request by dchaplinsky**

### CODE REVIEW

File: declarations_site/catalog/static/js/compare-charts.js [hunk 1]
```diff
@@ -0,0 +1,124 @@
+function makechart1() {
```
Comment (dchaplinsky) [inline declarations_site/catalog/static/js/compare-charts.js]: Can we have proper names here?

File: declarations_site/catalog/static/js/compare-charts.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    var ctx = document.getElementById("declRelatives").getContext("2d"),
```
Comment (dchaplinsky) [inline declarations_site/catalog/static/js/compare-charts.js]: ctx = $('#declRelatives').get(0).getContext("2d")
maybe?

File: declarations_site/catalog/static/js/compare-charts.js [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        jUrl = $('#declRelatives').data('jurl');
+
+    $.getJSON(jUrl, function(data) {
+        var barChartData = data;
+
+        window.myBar = new Chart(ctx, {
```
Comment (dchaplinsky) [inline declarations_site/catalog/static/js/compare-charts.js]: What is the point to expose this class to global namespace and later overwrite it in the method makechart1

File: declarations_site/catalog/static/js/compare-charts.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                            stepSize: 1
+                        }
+                    }]
+                }
+            }
+        });
+    });
+}
+
+$( document ).ready(function() {
```
Comment (dchaplinsky) [inline declarations_site/catalog/static/js/compare-charts.js]: $(function(){

})

File: declarations_site/catalog/static/js/compare-charts.js [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    console.log( "ready!" );
```
Comment (dchaplinsky) [inline declarations_site/catalog/static/js/compare-charts.js]: drop it

File: declarations_site/catalog/static/js/decl-compare.js [hunk 6]
```diff
@@ -0,0 +1,219 @@
+//read decl.ID from localstorage and populate compare list
+function getCompareList() {
+    var c = 0;
+
+    $.each(localStorage, function(key, value){
+        if (key.indexOf('declarationID-') >= 0) {
+            c++;
+            $('.decl-compare-toggle').removeClass('hidden');
+
+            var declarationID = key.substring(14);
```
Comment (dchaplinsky) [inline declarations_site/catalog/static/js/decl-compare.js:10]: What is 14? Some kind of magic number?

File: declarations_site/catalog/static/js/decl-compare.js [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+    //can compare
+    if (c > 1) {
+        $('.compare-list-action').removeClass('hidden');
+    }
+}
+
+function ClearCompareList() {
+    $.each(localStorage, function(key, value){
+        if (key.indexOf('declarationID-') >= 0) {
+            var declarationID = key.substring(14);
```
Comment (dchaplinsky) [inline declarations_site/catalog/static/js/decl-compare.js:38]: Make it a helper finally!

File: declarations_site/catalog/static/js/decl-compare.js [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+    //can compare
+    if (c > 1) {
+        $('.compare-list-action').removeClass('hidden');
+    }
+}
+
+function ClearCompareList() {
+    $.each(localStorage, function(key, value){
+        if (key.indexOf('declarationID-') >= 0) {
```
Comment (dchaplinsky) [inline declarations_site/catalog/static/js/decl-compare.js:37]: camelCase, underscores or dashes?

File: declarations_site/catalog/static/js/decl-compare.js [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+
+    var c = 0;
+    $.each(localStorage, function(key, value){
+        if (key.indexOf('declarationID-') >= 0) {
+            c++;
+        }
+    });
+
+    //add counter to badge
+    $('.decl-compare-toggle .badge').html(c);
```
Comment (dchaplinsky) [inline declarations_site/catalog/static/js/decl-compare.js:128]: Make 10 lines above a helper too, call it "update_number_of_declarations" or so.

DRY.

File: declarations_site/catalog/static/js/decl-compare.js [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+        $('.decl-compare-toggle').addClass('hidden');
+        $('.decl-compare-modal').modal('hide');
+        $('.compare-list-action').addClass('hidden');
+        $('.clear-compare-list').addClass('hidden');
+    }
+
+    //can't compare
+    if (c < 2) {
+        $('.compare-list-action').addClass('hidden');
+    }
```
Comment (dchaplinsky) [inline declarations_site/catalog/static/js/decl-compare.js:142]: Code above can go into the same helper. One piece of code would decide what to show and what to hide and you can run it from both event listeners

File: declarations_site/catalog/static/js/decl-compare.js [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+//add declaration 2 compare list by click from declaration-page
+$(document).on('click', '.declaration-page #page .add2compare-list', function(e){
+    var $container = $('.declaration-page #page'),
+        name = $container.find('#page-header span').text(),
+        desc = $container.find('.sub-header').prop('outerHTML'),
+        declarationID = $('#page').data('declid'),
+        url = '/declaration/' + declarationID,
+        itemHtml = '<div class="decl-item item col-sm-12  col-md-6 col-lg-4 grid-group-item nacp selected" itemscope="" itemtype="http://schema.org/Person"  data-declid="';
+
+    itemHtml = itemHtml + declarationID + '"><div class="box"><a href="#" class="add2compare-list" title="Додати у список порівнянь"><i class="fa fa-times" aria-hidden="true"></i>';
```
Comment (dchaplinsky) [inline declarations_site/catalog/static/js/decl-compare.js:162]: On declaration page you can render the same snippet as on search results page, but hide it with CSS.

This way you can control it from one point (jinja macro) and can get rid of this spaghetti

File: declarations_site/catalog/static/js/decl-compare.js [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+    itemHtml = itemHtml + '<i class="fa fa-plus" aria-hidden="true"></i><i class="fa fa-balance-scale" aria-hidden="true"></i></a><h3><a href="/declaration/'+ declarationID + '" itemprop="name">';
+    itemHtml = itemHtml +  name + '</a></h3>';
+    itemHtml = itemHtml + desc + '</div></div>';
+
+    if (localStorage.getItem("declarationID-" + declarationID) === null) {
```
Comment (dchaplinsky) [inline declarations_site/catalog/static/js/decl-compare.js]: Fuck. Make it a method or a helper. Call it "addDeclaration" or "add_declaration". Call it here and above

File: declarations_site/catalog/static/js/decl-compare.js [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+        localStorage.setItem("declarationID-" + declarationID, itemHtml);
+
+        $('.declaration-page #page .add2compare-list').addClass('active').html('У списку порівнянь');
+        $('.decl-compare-modal .modal-body').html($('.decl-compare-modal .modal-body').html() + itemHtml);
+
+        //show message to user
+        bootstrap_alert.warning('Декларацію додано до списку порівнянь', 'success', 2000);
+
+        var c = 0;
```
Comment (dchaplinsky) [inline declarations_site/catalog/static/js/decl-compare.js:177]: And call the helper mentioned three comments above here

File: declarations_site/catalog/static/js/decl-compare.js [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+    getCompareList();
+});
+
+$('.decl-compare-modal').on('shown.bs.modal', function () {
+    setColumnsHeights ('.decl-compare-modal .modal-body', '.item');
+});
+
+$( window ).on( 'resize', function () {
+    setColumnsHeights ('.decl-compare-modal.in .modal-body', '.item');
+});
```
Comment (dchaplinsky) [inline declarations_site/catalog/static/js/decl-compare.js:222]: I should mention, that we do not support comparison of old declarations (yet) and also change forms (as they cannot be compared).

File: declarations_site/catalog/static/js/main.js [hunk 15]
```diff
@@ -358,6 +358,17 @@ $(function() {
             });
     }
 
+    bootstrap_alert = function () {};
+    bootstrap_alert.warning = function (message, alert, timeout) {
```
Comment (dchaplinsky) [inline declarations_site/catalog/static/js/main.js:362]: Bad naming, you are using bootstrap_alert.warning to show success message.

**Review by dchaplinsky — COMMENTED** [review]
Do not merge yet

File: declarations_site/catalog/static/js/decl-compare.js (hunk 6 above)
Comment (ArialBlack) [reply declarations_site/catalog/static/js/decl-compare.js:10]: position at string

Comment (dchaplinsky) [reply declarations_site/catalog/static/js/decl-compare.js:10]: I do understand that, but:

1. Make it constant.
2. Explain it. You may use comments or you may calculate it, like this: my_magic_number = "my_string_prefix".length;

**Pull request merged**